feat(core): support fakeAsync.wrap() functionality across multiple test cases.

JiaLiPassion opened the pull request:
Close #40387

Enable global `fakeAsync()` feature by introducing hooks functions.
`fakeAsync.wrap({beforeEach?: () => void, afterEach?: () => void)`;

# Motivation

Consider the similar function from `jest.useFakeTimers()`, we can write test cases like this.

```
describe('test', () => {
   beforeAll(() => {
     jest.useFakeTimers();   
   });

   afterAll(() => {
     jest.clearAllTimers();
     jest.useRealTimers();
   });

   test('async test1', () => {
      setInterval(() => {}, 100);
      jest.advanceTimersByTime(100);
      expect(...);
   });

    test('async test2', () => {
      setInterval(() => {}, 200);
      jest.advanceTimersByTime(200);
      expect(...);
   });
});
```

So with `jest.useFakeTimers()`, all the following tests are automatically using the fake timers, and the developers can write the common `init`, `cleanup` logic in `beforeEach()/afterEach()`. 

But with `fakeAsync()`, we can not do the same thing for now.

For example, we have a Component like this.

```
@component({...})
export class AppComponent {
  timerId: number;
  ngOnInit() {
    this.timerId = setTimeout(() => {});
  }

  ngOnDestroy() {
    clearTimeout(this.timerId);
  }
}

And for now, we need to write test
like this.

describe('AppComponent test', () => {
  let fixture: ComponentFixture<AppComponent>;
  beforeEach(() => {
    ...
    fixture = TestBed.createComponent(AppComponent);
  });

  it('test case1', fakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
    fixture.destroy();
  }));

  it('test case2', fakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
    fixture.destroy();
  }));
});
```

We need to call `fixture.destroy()` inside each tests, since each it() use
it's own fakeAsync() and we need to clean up the timerId created in that
FakeAsyncZoneSpec. Otherwise the `fakeAsync()` will throw `there are 
still pending timers` error.


# Solution
So in this PR, there are two hooks functions are introduced. 

With the hook functions, we can write case in this way.
   
```
   describe('AppComponent test', () => {
     let fixture: ComponentFixture<AppComponent>;
   
     const fakeAsyncWithFixture = fakeAsync.wrap({
       beforeEach: () => {
         fixture = TestBed.createComponent(AppComponent);
         fixture.detectChanges();
       }
       afterEach: () => fixture.destroy();
     });
   
     it('test case1', fakeAsyncWithFixture(() => {
       // do some test with fixture
     }));
   
     it('test case2', fakeAsyncWithFixture(() => {
       // do some test with fixture
     }));
   });
   
   Also the `wrap()` function support nesting.
   
   describe('AppComponent test', () => {
     let fixture: ComponentFixture<AppComponent>;
   
     const fakeAsyncWithFixture = fakeAsync.wrap({
       beforeEach: () => {
         fixture = TestBed.createComponent(AppComponent);
         fixture.detectChanges();
       }
       afterEach: () => fixture.destroy();
     });
   
     it('test case1', fakeAsyncWithFixture(() => {
       // do some test with fixture
     }));
   
     it('test case2', fakeAsyncWithFixture(() => {
       // do some test with fixture
     }));
   
     describe('AppComponent sub test: auth test', () => {
       const fakeAsyncNested = fakeAsyncWithFeature.wrap({
         beforeEach: () => fixture.componentInstance.login(),
         afterEach: () => fixture.componentInstance.logout();
       });
   
       it('should show user info', () => {
         // do some test with fixture with authenticated user.
       });
     });
   });
```

This feature will be useful when we want to do some global cleanup
in `afterEach()` when the component have some `async` tasks
(especially `setInterval`) running by 3rd party library.

JiaLiPassion committed:
feat(zone.js): add `fakeAsync.wrap()` function to support `beforeEach()/afterEach()` hooks.

Close #40387

Provide the new `fakeAsync.wrap()` function with the `beforeEach()/afterEach()` hooks.

Given the AppComponent:

@Component({...})
export class AppComponent {
  timerId: number;
  ngOnInit() {
    this.timerId = setTimeout(() => {});
  }

  ngOnDestroy() {
    clearTimeout(this.timerId);
  }
}

And without hook functions, we need to write test
like this.

describe('AppComponent test', () => {
  let fixture: ComponentFixture<AppComponent>;
  beforeEach(() => {
    ...
    fixture = TestBed.createComponent(AppComponent);
  });

  it('test case1', fakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
    fixture.destroy();
  }));

  it('test case2', fakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
    fixture.destroy();
  }));
});

We need to call `fixture.destroy()` inside each tests, since each `it()` use
it's own fakeAsync() and we need to clean up the timerId created in that
FakeAsyncZone.

With the hook functions, we can write case in this way.

describe('AppComponent test', () => {
  let fixture: ComponentFixture<AppComponent>;

  const fakeAsyncWithFixture = fakeAsync.wrap({
    beforeEach: () => {
      fixture = TestBed.createComponent(AppComponent);
      fixture.detectChanges();
    }
    afterEach: () => fixture.destroy();
  });

  it('test case1', fakeAsyncWithFixture(() => {
    // do some test with fixture
  }));

  it('test case2', fakeAsyncWithFixture(() => {
    // do some test with fixture
  }));
});

Also the wrap() function support nesting.

describe('AppComponent test', () => {
  let fixture: ComponentFixture<AppComponent>;

  const fakeAsyncWithFixture = fakeAsync.wrap({
    beforeEach: () => {
      fixture = TestBed.createComponent(AppComponent);
      fixture.detectChanges();
    }
    afterEach: () => fixture.destroy();
  });

  it('test case1', fakeAsyncWithFixture(() => {
    // do some test with fixture
  }));

  it('test case2', fakeAsyncWithFixture(() => {
    // do some test with fixture
  }));

  describe('AppComponent sub test: auth test', () => {
    const fakeAsyncNested = fakeAsyncWithFeature.wrap({
      beforeEach: () => fixture.componentInstance.login(),
      afterEach: () => fixture.componentInstance.logout();
    });

    it('should show user info', () => {
      // do some test with fixture with authenticated user.
    });
  });
});

This feature is useful when we want to do some common initial/cleanup when the component has some `async`
tasks (especially `setInterval`) running by 3rd party library.

mhevery commented:
What happens when developer forgets to invoke `endFakeAsync()`? Is there a sane failure mode?

An alternative approach may be something as shown below. Have you consider it? What do you think are advantages/disadvantages of the two approaches?

```
describe('AppComponent test', () => {
  let fixture: ComponentFixture<AppComponent>;
  const withFixtureFakeAsync = fakeAsync.bind({
    before:  () => fixture = TestBed.createComponent(AppComponent),
    after: () => fixture.destroy()
  });

  it('test case1', withFixtureFakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
  }));

  it('test case2', withFixtureFakeAsync(() => {
    fixture.detectChanges();
    // do some test with fixture
  }));
});

```

JiaLiPassion commented:
@mhevery, yeah, I like your idea, 
1. This is much safer, the users don't need to worry about forgetting to call `endFakeAsync()`.
2. And the user will know that their tests are under `fakeAsync()`, also in the same describe, they can write `fakeAsync()`, `async()`, `normal` test together.

I will update the PR

mhevery commented on aio/content/examples/testing/src/app/auth/auth.component.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    .createComponent(AuthComponent);
+  });
+
+  const withFeatureFakeAsync = fakeAsync.bind({
```
NIT: I would flip the name to `fakeAsyncWithFixture`

JiaLiPassion replied:
Got it, thanks.

mhevery commented on aio/content/examples/testing/src/app/auth/auth.component.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    expect(fixture.componentInstance.result).toEqual('work1 done after auth with token user1 token1');
+  }));
+
+  it('shouldWork2 should work correctly', withFeatureFakeAsync(() => {
```
It would be nice to move this test as a subtest of `describe` and add: 
```
const fakeAsyncWithSomethingFixtrue = fakeAsyncWithFixture.bind({
  before: ...,
  after: ...,
});
```

This way we can demonstrate that we can nest the bindings. (Also I think your current implementation does not support this feature)

JiaLiPassion replied:
Yes, you are right, I have updated the implementation to support the nesting wrap, and also added the test case.

mhevery commented on aio/content/examples/testing/src/app/auth/auth.component.spec.ts:
```diff
@@ -0,0 +1,41 @@
+import { ComponentFixture, fakeAsync, TestBed, tick } from '@angular/core/testing';
```
I think you still need to have a file header which includes copyright.

mhevery commented on aio/content/examples/testing/src/app/auth/auth.component.ts:
```diff
@@ -0,0 +1,35 @@
+// #docregion
```
I think you still need to have a file header which includes copyright.

JiaLiPassion replied:
Got it, I added the header.

mhevery commented on aio/content/examples/testing/src/app/auth/auth.component.spec.ts:
```diff
@@ -0,0 +1,41 @@
+    .createComponent(AuthComponent);
+  });
+
+  const withFeatureFakeAsync = fakeAsync.bind({
```
Maybe also add a comment describing what you are trying to demonstrate.

JiaLiPassion replied:
Got it, I added the comment here.

mhevery commented on aio/content/examples/testing/src/app/model/auth.service.ts:
```diff
@@ -0,0 +1,35 @@
+import { Injectable } from '@angular/core';
```
I think you still need to have a file header which includes copyright.

JiaLiPassion replied:
Got it.

mhevery commented on goldens/public-api/core/testing/testing.d.ts:
```diff
@@ -29,6 +29,12 @@ export declare const ComponentFixtureNoNgZone: InjectionToken<boolean[]>;
 export declare function discardPeriodicTasks(): void;
 
 export declare function fakeAsync(fn: Function): (...args: any[]) => any;
+export declare namespace fakeAsync {
```
This should be typed as (delete `function fakeAsync`):
```suggestion
interface FakeAsync {
  // docs here
  (fn: Function): (...args: any[]) =>any;
  
  // docs here
  wrap({
    before?: ()=>void
    after?: ()=>void
  }): FakeAsync;
}
```

JiaLiPassion replied:
Got it, this resolves the typescript parsing failure issue in `aio`, thank you!

mhevery commented on packages/core/test/fake_async_spec.ts:
```diff
@@ -432,4 +432,67 @@ const ProxyZoneSpec: {assertPresent: () => void} = (Zone as any)['ProxyZoneSpec'
+      });
+
+      it('test App doSomeWork1', fakeAsyncWithHook(() => {
+           app.doSomeWork1();
```
missing test for nested `bind` calls. `fakeAsync.bind(...).bind(...)`

JiaLiPassion replied:
Got it, I added the nesting test cases.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -51,6 +51,78 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
+ * Bind `before` and `after` hooks to `fakeAsync()` to allow the user to invoke common `initial` and
+ * `cleanup` logic in the same scope of the test body.
+ *
+ * For example, we have a Component like this.
```
```suggestion
 * Given `AppComponent`:
 ```

JiaLiPassion replied:
Got it, updated, thanks

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -51,6 +51,78 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
+ *   }));
+ * });
+ *
+ * Here we need to call `fixture.destroy()` inside each test, since each it() is in its own
```
symbols such as `it()` should be back-ticked

JiaLiPassion replied:
Got it, I added the back-ticked for all functions.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -51,6 +51,78 @@ export function fakeAsync(fn: Function): (...args: any[]) => any {
+ *     // do some test with fixture
+ *   }));
+ *
+ *   it('test case2', withFeatureFakeAsync(() => {
```
need example of nested bind.

JiaLiPassion replied:
Got it, I added the example of nested wrap.

mhevery commented on packages/zone.js/lib/jasmine/jasmine.ts:
```diff
@@ -195,7 +195,7 @@ Zone.__load_patch('jasmine', (global: any, Zone: ZoneType, api: _ZonePrivate) =>
     const testProxyZone = queueRunner.testProxyZone!;
     let lastDelegate;
-    if (isClockInstalled && enableAutoFakeAsyncWhenClockPatched) {
+    if ((isClockInstalled && enableAutoFakeAsyncWhenClockPatched)) {
```
revert?

JiaLiPassion replied:
Got it, I reverted it.

mhevery commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -731,6 +731,89 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
     getProxyZoneSpec() && getProxyZoneSpec().assertPresent().resetDelegate();
   }
 
+  /**
```
I am confused. What is the relationship between this file and packages/core/testing/src/fake_async.ts? Seems like we are duplicating code?

JiaLiPassion replied:
Yes, some code are duplicate here for some legacy reasons, I created another PR to cleanup the code https://github.com/angular/angular/pull/39941

mhevery commented on packages/zone.js/lib/zone-spec/fake-async-test.ts:
```diff
@@ -802,6 +885,8 @@ Zone.__load_patch('fakeasync', (global: any, Zone: ZoneType, api: _ZonePrivate)
     return fakeAsyncFn;
   }
 
+  fakeAsync.bind = bind;
```
I was discussing this with @petebacondarwin and we have some suggestions. (I know I am the one who suggested we use `bind` but I really did not think through the implications.

- using `fakeAsync.bind` is problematic because `fakeAsync` is a `Function` which already has `bind` method. So what we are doing here is changing the `bind` method signature and semantics. We should not do that.

Because of the above reasons we should do something else. (Again sorry for suggesting this in first place)

Our suggestion is to use `fakeAsync.wrap(...)` which could be chained as `fakeAsync.wrap(...).warp(...)`

JiaLiPassion replied:
Got it, I updated the method name to `wrap()` and support nesting.

petebacondarwin commented:
By the way, should the hooks be `beforeEach` and `afterEach` to be more inline with the jasmine API?

JiaLiPassion commented:
@petebacondarwin , yes, you are right, `beforeEach()` and `afterEach()` are better names, I updated the PR, thank you.

mary-poppins commented:
You can preview fe770f6 at https://pr40611-fe770f6.ngbuilds.io/.

mary-poppins commented:
You can preview 0f8d2ac at https://pr40611-0f8d2ac.ngbuilds.io/.

mary-poppins commented:
You can preview 153891e at https://pr40611-153891e.ngbuilds.io/.

mary-poppins commented:
You can preview 4382afc at https://pr40611-4382afc.ngbuilds.io/.

mary-poppins commented:
You can preview 925d34a at https://pr40611-925d34a.ngbuilds.io/.

mary-poppins commented:
You can preview 6df66fb at https://pr40611-6df66fb.ngbuilds.io/.

alfaproject commented:
I see that you wrote `fakeAsyncWithFeature` in quite a few places but don't you mean `fakeAsyncWithFixture`?

mary-poppins commented:
You can preview 4d27a26 at https://pr40611-4d27a26.ngbuilds.io/.

JiaLiPassion commented:
@alfaproject , thank you, I have updated the PR.

mary-poppins commented:
You can preview 56845be at https://pr40611-56845be.ngbuilds.io/.

mary-poppins commented:
You can preview b721942 at https://pr40611-b721942.ngbuilds.io/.

mhevery commented on goldens/public-api/core/testing/testing.d.ts:
```diff
@@ -28,7 +28,7 @@ export declare const ComponentFixtureNoNgZone: InjectionToken<boolean[]>;
 export declare function discardPeriodicTasks(): void;
 
-export declare function fakeAsync(fn: Function): (...args: any[]) => any;
+export declare const fakeAsync: FakeAsync;
```
HMM, I see that you are exporting `fakeAsync` but not `FakeAsync` can you fix that? Also `FakeAsync` because it is public API will need public documentation.

JiaLiPassion replied:
Sure, I exposed the API. Thank you.

mhevery commented on packages/core/testing/src/fake_async.ts:
```diff
@@ -25,6 +25,129 @@ export function resetFakeAsyncZone(): void {
   throw new Error(fakeAsyncTestModuleNotLoadedErrorMessage);
 }
 
+interface FakeAsync {
```
Please add public docs, and export it.

mhevery commented:
I also did some fixes here: f64813dc5d

mary-poppins commented:
You can preview 0e6d5cd at https://pr40611-0e6d5cd.ngbuilds.io/.

JiaLiPassion committed:
feat(core): add fakeAsync.wrap() interface to core/testing

`zone.js` provides a new API `fakeAsync.wrap()`
to support `hooks` of `fakeAsync()` to allow user to write common `begin`/'after`
logic. This commit add the new `wrap()` API to `@angular/core/testing`.

dylhunn reviewed (dismissed):
reviewed-for: public-api

atscott commented:
Hi @JiaLiPassion, it looks like the issue you linked was closed by https://github.com/angular/angular/commit/0652b29f6288c7150e1ce143fdd1d3e509292c9d. Is this still necessary?

That question aside, I'm not sure I see why this specific thing should be a feature provided by the `core`/`zone`/`fakeAsync` API. How is it different from this:

```
  describe('AppComponent test', () => {
     let fixture: ComponentFixture<AppComponent>;
   
     beforeEach(() => {
       fixture = TestBed.createComponent(AppComponent);
       fixture.detectChanges();
     });
     afterEach(() => fixture.destroy());
   
     it('test case1', fakeAsync(() => {
       // do some test with fixture
     }));
   
     it('test case2', fakeAsync(() => {
       // do some test with fixture
     }));
   });
```

The above is actually shorter than your example using this new feature.

JiaLiPassion commented:
@atscott thank you for the review.
The reason we need a new `wrap` API is to make sure `before/after` hook can run in the same `fakeAsync()` zone with the `it()` test case.
If we write case like this,
```
@component({...})
export class AppComponent {
  timerId: number;
  ngOnInit() {
    this.timerId = setTimeout(() => {});
  }

  ngOnDestroy() {
    clearTimeout(this.timerId);
  }
}
  describe('AppComponent test', () => {
     let fixture: ComponentFixture<AppComponent>;
   
     beforeEach(() => {
       fixture = TestBed.createComponent(AppComponent);
       fixture.detectChanges();
     });
     afterEach(() => fixture.destroy());
   
     it('test case1', fakeAsync(() => {
       // do some test with fixture
      tick();
     }));
   
     it('test case2', fakeAsync(() => {
       // do some test with fixture
     }));
   });
```

the `setTimeout` in `beforeEach()` will not in the `fakeAsync()` and will run asynchronize, so the `it('test case1')` and `it('test case2')` will run before the `setTimeout callback` executed, and calling `tick()` in `it()` will not trigger the `setTimeout` in the `beforeEach()` which may not match the user's expection.

atscott commented:
@JiaLiPassion Okay, I think that makes sense. Could you please update the documentation to describe this better? The way it is now, it seems to imply what I wrote should work fine. "without hook functions, we need to write test like this" Unless I'm missing something, the example after that sentence is exactly what I wrote in my comment.

Also, I'm still curious about https://github.com/angular/angular/commit/0652b29f6288c7150e1ce143fdd1d3e509292c9d and if it makes this PR obsolete. If not, the commit message should be changed to reflect that this solves a different issue (one that's not closed).

JiaLiPassion commented:
@atscott , got it, I will update the commit message, and about the `aio` doc, in the previous commit, the `aio document` and the `sample code` are also be updated, but it seems the `aio` project reference the `13.2.2` version instead of the `local` version, so the sample code will not be compiled since the `new wrap()` API.

JiaLiPassion committed:
docs(core): add `fakeAsync.wrap()` docs into aio.

1. add `fakeAsync.wrap()` example code.
1. add `fakeAsync.wrap()` doc in the `testing-components-scenarios.md`
doc.

atscott commented:
@JiaLiPassion I've moved this back to a draft since it doesn't seem ready for review
